### PR TITLE
add double grad compute for batch norm

### PR DESCRIPTION
**Pull request by ceci3**
### PR types

Others

### PR changes

OPs

### Describe

add double grad compute for batch norm

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_imperative_double_grad.py [hunk 1]
```diff
@@ -346,7 +346,7 @@ def raise_no_grad_op(self):
         with fluid.dygraph.guard():
             x = fluid.layers.ones(shape=[2, 3, 2, 2], dtype='float32')
             x.stop_gradient = False
-            y = paddle.fluid.layers.batch_norm(x)
+            y = paddle.fluid.layers.group_norm(x, groups=1)
```
Comment (ceci3) [inline python/paddle/fluid/tests/unittests/test_imperative_double_grad.py:349]: 这个单测测试的是没有double grad的op的是否抛出异常，这个pr支持了batch norm的double grad， 所以把这个单测的batch_norm 换成group_norm

File: paddle/fluid/operators/norm_utils.cu.h [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+#include "paddle/fluid/platform/cudnn_helper.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+using DataLayout = framework::DataLayout;
+
+// scale shape == mean shape
+template <typename T, int BlockDim, framework::DataLayout layout>
```
Comment (heavengate) [inline paddle/fluid/operators/norm_utils.cu.h:50]: 注释里给一下计算公式？

Comment (ceci3) [reply paddle/fluid/operators/norm_utils.cu.h:50]: 已添加~

File: paddle/fluid/operators/batch_norm_op.h [hunk 3]
```diff
@@ -103,6 +103,44 @@ inline void TransToChannelFirst(const framework::ExecutionContext& context,
   }
 }
 
+template <typename DeviceContext, typename T>
+inline void ResizeToChannelLast(const framework::ExecutionContext& context,
+                                const Tensor* input,
+                                Tensor* transformed_input) {
+  int dim = input->dims().size() - 2;
+  if (dim == 3) {
+    // input
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/batch_norm_op.h]: 这个注释是为了说明什么？

File: python/paddle/fluid/tests/unittests/test_norm_nn_grad.py [hunk 4]
```diff
@@ -68,5 +68,67 @@ def func(self, place):
...
+            eps = 0.005
+            atol = 1e-4
+            x = layers.create_parameter(dtype=dtype, shape=self.shape, name='x')
+            z = fluid.layers.batch_norm(
+                input=x,
+                data_layout=self.data_layout,
+                use_global_stats=self.use_global_stats)
+            x_arr = np.random.uniform(-1, 1, self.shape).astype(dtype)
+            gradient_checker.double_grad_check(
+                [x], z, x_init=x_arr, atol=atol, place=place, eps=eps)
```
Comment (wanghaoshuang) [inline python/paddle/fluid/tests/unittests/test_norm_nn_grad.py:95]: 为什么不用默认的atol呢？

**Review by wanghaoshuang — DISMISSED** [review]

File: paddle/fluid/operators/batch_norm_op.h (hunk 3 above)
Comment (ceci3) [reply paddle/fluid/operators/batch_norm_op.h]: 已删除~

File: python/paddle/fluid/tests/unittests/test_norm_nn_grad.py (hunk 4 above)
Comment (ceci3) [reply python/paddle/fluid/tests/unittests/test_norm_nn_grad.py:95]: norm的二阶梯度计算公式里比较复杂，容易导致单测里解析结果和数值结果的值不能完全对上，竞品单测针对norm的二阶梯度也单独设置了atol~

**Review by wanghaoshuang — APPROVED** [review]

**Review by zhiqiu — APPROVED** [review]
LGTM for `fluid.layers.instance_norm` without `core.ops`, since it is used for static-graph only.
LGTM for using `ShareDataWith` between temoral tensor and input.

**Review by heavengate — APPROVED** [review]

**Review by zhangting2020 — APPROVED** [review]
LGTM

**Review by raindrops2sea — APPROVED** [review]
